include/common/coredump.h

```diff
@@ -0,0 +1,287 @@
+
+    AST::MemorySection MemSec;
+    const auto *CurrentInstance = StackMgr.getModule();
+    auto MemInstances = CurrentInstance->getOwnedMemoryInstances();
+    std::cout << MemInstances[0]->getPageSize();
+    // In the current version of WebAssembly, at most one memory may be defined
+    // or imported in a single module, and all constructs implicitly reference
+    // this memory.
+    // std::cout << MemInstances[0]->getMemoryType().getLimit().getMin();
+    // auto &Content = MemSec.getContent();
+    // TODO why is this setting max and min to the same value?
+    // for (auto &Iter : MemInstances) {
+    //   AST::MemoryType MemType;
+    //   auto Entry = Iter->getMemoryType();
+    //   if (Entry.getLimit().hasMax()) {
+    //     spdlog::warn("Has max");
+    //     auto Max = Entry.getLimit().getMax();
+    //     MemType.getLimit().setMax(Max);
+    //   }
+    //   MemType.getLimit().setMin(0);
+    //   Content.push_back(MemType);
+    // }
+
+    return MemSec;
+  }
+
+  // Data section
+  // TODO malformed section
+  AST::DataSection collectDataSection(Runtime::StackManager &StackMgr) {
+    spdlog::info("Collecting Memories");
+
+    AST::DataSection Datasec;
+
+    const auto *CurrentInstance = StackMgr.getModule();
+    auto &DataInstances = CurrentInstance->getOwnedDataInstances();
+
+    AST::DataSegment Seg;
+    auto &Content = Seg.getData();
+
+    for (auto &Iter : DataInstances) {
+      Content.insert(Content.end(), Iter->getData().begin(),
+                     Iter->getData().end());
+    }
+    if (Content.size() == 0) {
+      std::cout << "empty\n";
+      std::cout << Content.max_size();
+      Content.insert(Seg.getData().end(), 0x00);
+    }
+    std::cout << Content.size();
+    Datasec.getContent().push_back(Seg);
+    return Datasec;
+  }
+
+  // corestack   ::= customsec(thread-info vec(frame))
+  // thread-info ::= 0x0 thread-name:name
+  // frame       ::= 0x0 funcidx:u32 codeoffset:u32
+  //                     locals:vec(value) stack:vec(value)
+  AST::CustomSection collectCoreStack(Runtime::StackManager &StackMgr) {
+    spdlog::info("Constructing CoreStack");
+
+    AST::CustomSection CoreStack;
+    CoreStack.setName("corestack");
+
+    auto Frames = StackMgr.getAllFrames();
+
+    // TODO get thread info
+    auto &Content = CoreStack.getContent();
+    for (auto &Item : Frames) {
+      if (Item.Module == nullptr) // guard
+        continue;
+
+      auto Funcidx = Item.From->getSourceIndex(); // function index
+      // TODO which to use?
+      //  auto Funcidx = Item.From->getTargetIndex(); // function index
+      auto Codeoffset = Item.From->getOffset();
+      // auto Codeoffset = 0; // only if codeoffset is unknown?
+      auto LocalsN = Item.Locals;
+
+      auto Locals = StackMgr.getRangeSpan(Item.VPos, LocalsN + Item.VPos);
+      auto Stack = StackMgr.getTopSpan(StackMgr.size() -
+                                       LocalsN); // TODO is this correct?
```

Comment:
You will need to use the relationship between N-1, N, and N+1 to retrieve the locals and stacks.

```
Frames |...|        N-1 frame          |        N frame        |         N+1 frame         | ... |
Stack  |...| Locals(N-1) | Stacks(N-1) | Locals(N) | Stacks(N) | Locals(N+1) | Stacks(N+1) | ... |
       |...|              ^            |            ^          |              ^            | ... |
       |...|              | VPos(N-1)  |            | VPos(N)  |              | VPos(N+1)  | ... |
```

Reply:
While collecting these values is there any way to also get the binary encoding of their types? All values are of type `ValVariant` I think for these at least if I write code to match for the following types it shouldn't really be an issue.

value ::= 0x01       => ∅
        | 0x7F n:i32 => n
        | 0x7E n:i64 => n
        | 0x7D n:f32 => n
        | 0x7C n:f64 => n

Reply:
Yeah, i don't think v128 is supported in wasmgdb. Regarding the stack trace as well: Funcidx maps perfectly to `targetIndex` for a frame's `From` iterator.

Reply:
Does the arity play a role in how many values are in the value stack as well?

Reply:
> Does the arity play a role in how many values are in the value stack as well? Is there a simpler way to get both locals and stack values for a particular frame than this?

No. Arity means the number of the return values. So you cannot use it for calculating the range of the stack. You should use VPos and LocalsN to calculate just like I mentioned above.

Reply:
Yep got it, the locals and stack variables have been captured correctly now. The issue is now in the binary encoding within the coredump which I am figuring out.